ThunkLibs/libwayland-client/Guest.cpp

```diff
@@ -0,0 +1,235 @@
+/*
+$info$
+tags: thunklibs|wayland-client
+$end_info$
+*/
+
+#include <wayland-util.h>
+#include <wayland-client.h>
+
+// These must be re-declared with an initializer here, since they don't get exported otherwise
+// NOTE: The initializers for these must be fetched from the host Wayland library, however
+//       we can't control how these symbols are loaded since they are global const objects.
+//       LD puts them in the application rodata section and ignores any nontrivial library-provided
+//       initializers. There is a workaround to enable late initialization anyway in OnInit.
+// NOTE: We only need to do this for interfaces exported by libwayland-client itself. Interfaces
+//       defined by external libraries work fine.
+extern "C" const wl_interface wl_output_interface {};
+extern "C" const wl_interface wl_shm_pool_interface {};
+extern "C" const wl_interface wl_pointer_interface {};
+extern "C" const wl_interface wl_compositor_interface {};
+extern "C" const wl_interface wl_shm_interface {};
+extern "C" const wl_interface wl_registry_interface {};
+extern "C" const wl_interface wl_buffer_interface {};
+extern "C" const wl_interface wl_seat_interface {};
+extern "C" const wl_interface wl_surface_interface {};
+extern "C" const wl_interface wl_keyboard_interface {};
+extern "C" const wl_interface wl_callback_interface {};
+
+#include <array>
+#include <charconv>
+#include <cstdio>
+#include <cstdarg>
+#include <cstring>
+#include <string>
+
+#include "common/Guest.h"
+
+#include "thunkgen_guest_libwayland-client.inl"
+
+struct wl_proxy_private {
+  wl_interface* interface;
+  // Other data members omitted
+};
+
+template<char> struct ArgType;
+template<> struct ArgType<'s'> { using type = const char*; };
+template<> struct ArgType<'u'> { using type = uint32_t; };
+template<> struct ArgType<'i'> { using type = int32_t; };
+template<> struct ArgType<'o'> { using type = wl_proxy*; };
+template<> struct ArgType<'a'> { using type = wl_array*; };
+template<> struct ArgType<'f'> { using type = wl_fixed_t; };
+template<> struct ArgType<'h'> { using type = int32_t; }; // fd?
+
+template<char... Signature>
+static void* WaylandAllocateHostTrampolineForGuestListener(void (*callback)()) {
+  using cb = void(void*, wl_proxy*, typename ArgType<Signature>::type...);
+  return (void*)AllocateHostTrampolineForGuestFunction((cb*)callback);
+}
+
+#define WL_CLOSURE_MAX_ARGS 20
+
+extern "C" int wl_proxy_add_listener(struct wl_proxy *proxy,
+      void (**callback)(void), void *data) {
+  auto interface = ((wl_proxy_private*)proxy)->interface;
+
+  // NOTE: This pointer must remain valid past the return of this function.
+  // TODO: What's a good way to manage the lifetime of this?
```

Comment:
Could probably create a map of proxies that have listeners attached and destroy the allocation on `wl_proxy_destroy`?

Reply:
Yes, was a bit worried about potential other ways to destroy proxies and about multiple listeners being set up. But proxy wrappers don't get their own set of listener callbacks, and `wl_proxy_add_listener` can't be called more than once per proxy. So fixing this was fairly simple after all :)